metal3-scaling-experiments/README.md

```diff
@@ -51,156 +53,155 @@ Caveats:
 - The Kubeadm control plane provider is trying to reach the (fake) static pods to check certificate expiration.
   To try to mitigate this, we attempted to set the expiration annotation on the KubeadmConfig, but unfortunately this caused some KCPs to start rollouts. It is unclear what is causing this.
 
+### Version 2
+
+1. Create a kind cluster, initialize Cluster API and Metal3 with BMO in test-mode. Create the multi-tenant ETCD instance.
+   `./setup-scaling-experiment-v2.sh`
+2. Add more workload clusters:
+   `./create-clusters-v2.sh <number-of-clusters>`
+
+To solve the issues with version 1, this setup creates one API server per workload cluster, all backed by one multi-tenant etcd instance.
+This simulation is much more realistic as each cluster has its own API and do not see each others Nodes or Pods.
+However, running one API server per cluster takes a lot of resources.
+Each one needs roughly 200 Mi memory, so scaling to 1000 clusters would require at least 200 Gi memory just for the API servers.
+
+### Version 3 - External backing clusters
+
+- Setup a (kind) cluster on an external machine and expose it in some way so that the test machine can reach it.
+  See the `ansible` folder for a useful playbook that can help with this.
+- `./setup-scaling-experiment-v3.sh` on the test machine
+- For each backing cluster:
+   - Set KUBECONFIG to point to the external cluster and run `./setup-backing-cluster.sh`.
+     This will create one multi-tenant etcd in the backing cluster
+- `./create-clusters-v3.sh <number-of-clusters> <kubeconfig> <start>`
+   This will create `<number-of-clusters>` with the backing cluster that `<kubeconfig>` points to starting with cluster number `<start>`. (The starting number is to avoid collisions when using multiple backing clusters, since all cluster objecs live in the management cluster.) E.g. `./create-clusters-v3.sh 100 kubeconfig-backing-2.yaml 101`.
+
+For scaling to 1000, I have done the following also:
+
+- Set all concurrency flags to 100 (default is 10) for the CAPI controllers.
+- Set a higher rate limit (200 QPS and 300 burst) for all the CAPI controllers (default is 20 QPS and 30 burst).
+
```

Comment:
Could you please highlight here, that approx how many backing cluster would the user need for the 1000 simulated clusters ? I would guess based on your description and this https://etcd.io/docs/v3.3/dev-guide/limit/ that you would need 25 multi-tenant etcd , is it true our would be fewer etcd enough ?

Reply:
Good point! I'll add what I used (10 backing clusters with 100 clusters each).

Reply:
Done